api/server.go

```diff
@@ -93,6 +87,10 @@ func (srv *Server) Serve() error {
 	// closed, via either the Close method or the signal handling above.
 	// Closing the listener will result in the benign error handled below.
 	err := srv.apiServer.Serve(srv.listener)
+
+	// Wait for srv.Close to finish before returning.
+	srv.wg.Wait()
```

Comment:
You call `Wait()` before handling the error? Also, why is the waiting performed in the `Serve()` thread instead of the `Close()` thread?

I'm not following the shutdown design very well here. `Close()` needs to block until all processes have completed, which would include the `Serve()` process.